### ISSUE TITLE
duckdns can't obtain cert after upgrade to 1.12.5 

### ISSUE THREAD

**Issue by kricha** [home-assistant/addons#1869]
## The problem

After upgrading addon to 1.12.5 version, it can't obtain certificate

## Environment

- Add-on with the issue: DuckDNS
- Add-on release with the issue:  
- Last working add-on release (if known): 
- Operating environment (OS/Supervised):  Home Assistant 2021.2.3

## Problem-relevant configuration

```yaml
lets_encrypt:
  accept_terms: true
  certfile: fullchain.pem
  keyfile: privkey.pem
token: duckdns_token
domains:
  - myownmane.duckdns.org
aliases:
  - domain: owndomain.com
    alias: myownmane.duckdns.org
seconds: 300
```

## Traceback/Error logs

```txt
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
# INFO: Using main config file /data/workdir/config
+ Account already registered!
[15:27:13] INFO: OK
185.124.168.108
NOCHANGE
[15:27:15] INFO: Renew certificate for domains: myownmane.duckdns.org and aliases: 
owndomain.com
# INFO: Using main config file /data/workdir/config
Processing owndomain.com with alternative names: myownmane.duckdns.org
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 2 authorizations URLs from the CA
 + Handling authorization for owndomain.com
 + Handling authorization for myownmane.duckdns.org
 + 2 pending challenge(s)
 + Deploying challenge tokens...
OKOK + Responding to challenge for owndomain.com authorization...
 + Cleaning challenge tokens...
OKOK + Challenge validation has failed :(
ERROR: Challenge is invalid! (returned: invalid) (result: {
  "type": "dns-01",
  "status": "invalid",
  "error": {
    "type": "urn:ietf:params:acme:error:unauthorized",
    "detail": "Incorrect TXT record \"_JSFK0dRBHcg1klisURl0aHdq1aCiZ_4imd8ZHupHhI\" found at _acme-challenge.owndomain.com",
    "status": 403
  },
  "url": "https://acme-v02.api.letsencrypt.org/acme/chall-v3/someints/somealphas",
  "token": "qNK-toen"
})
```

## Additional information
<img width="1205" alt="CleanShot 2021-02-17 at 15 48 53@2x" src="https://user-images.githubusercontent.com/4619899/108213455-b8a6a100-7137-11eb-9c4e-d67eee8fbb19.png">

All works great before upgrade

**Comment by matlimatli** [comment]
I have exactly the same issue. Upgraded to 1.12.5, modified the config as required by #1785, and now get the same negative result as above (Challenge is invalid). I am not sure which was the last working version. I also tried reinstalling the plugin.

If I remove the aliases, the certificate request works fine for the duckdns domain.

**Comment by OndraM** [comment]
Same behavior here. I reconfigured DuckDNS settings following #1785 according to [updated documentation](https://github.com/RyanMeulenkamp/addons/blob/master/duckdns/DOCS.md#option-aliases-optional), so that `domains` contain only duckdns host and `aliases` adds my own 3rd level domain.

I also made sure CNAME DNS record is as stated by the documentation.

As a result, I'm getting `Incorrect TXT record` as well with addon unable to reissue certificate for my domain with the same message as in the issue above.

**Comment by rruizGit** [comment]
Same exact issue with my setup. Didn't know about the change with #1785 so my cert expired. I updated as discussed above and got exactly the same failure.

**Comment by andyeskridge** [comment]
I am also hitting this issue. @rruizGit what do you mean forwarding your 'A' record? Did you just create a new A record with the current IP address of your home assistant?

**Comment by andyeskridge** [comment]
Alright, well I worked around this by turning off the certs in duckdns and installing the let's encrypt addon and using that to generate the certificates for me.

**Comment by rruizGit** [comment]
Sorry, crazy things going on in my life. All I did was go to my Domain/DNS host and have it point my 'A' record to the IP address of my DuckDNS IP. But truthfully, not sure if that fixed the issue or not. It could very well be that I got it to work with the previous version and then upgraded. For all I know when that cert expires the latest code could fail again. 
Good to know that I could always move to just using "Let's Encrypt" directly.

**Comment by Orgjvr** [comment]
I am also hitting this issue. What is happening is that the addon is requesting challenges for both the alias(es) as well as the domain(s). The second challenge is overwriting the first challenge before the challenge validation is taking place, thus resulting in a failed validation. 

As a temporary workaround 
1. Remove all the aliases from the config and let it just validate the Duckdns domains. This completes successfully.
2. Put back the aliases into the config and let it re-validate all the domains. As the Duckdns domains are already validated, it will only deploy challenges to the Alias domains. This time it will complete successfully (if your alias domains point to different duckdns domains! In case more than one Alias point to the same Duckdns domain, you need to split step 2 into multiple steps and let it only validate one of the duplicate aliases at a time).

**Comment by sigo** [comment]
Same here. Since update cert cannot be renew.

**Comment by sigo** [comment]
Spent too much time on this.

Duck DNS addon cannot properly work with renewing cert for domain with aliases at this moment. I'm going to Duck DNS (for DDNS service) with Let's encrypt addon combo.

It is required to set auto starting Let's encrypt every day/week or so.

**Comment by rruizGit** [comment]
@sigo, do you mind sharing how you setup Let's Encrypt with aliases to get this working? I'm going through this yet again.

**Comment by sigo** [comment]
@rruizGit sure, but the configs isn't rocket science. I'll provide quick guide. Config details depends on DNS provider ([Supported DNS providers](https://github.com/home-assistant/addons/blob/master/letsencrypt/DOCS.md#supported-dns-providers)).

[Let's Encrypt documentation.](https://github.com/home-assistant/addons/blob/master/letsencrypt/DOCS.md) It is a bit lengthy due to many supported DNS providers. Worth reading.

1. Have `Duck DNS` and `Let's Encrypt` plugin installed.
2. Have only one DNS rule - `CNAME <your-domain> -> <domain>.duckdns.org`. Can be proxied on Cloudflare. You can safely delete any other created for `DuckDNS`.
2. `DuckDNS` config
```yml
lets_encrypt:
  accept_terms: false
  certfile: fullchain.pem
  keyfile: privkey.pem
token: <SECRET>
domains:
  - <SECRET>.duckdns.org
aliases: []
seconds: 300
```
3. `Let's Encrypt` config mostly depends on DNS provider. I use Cloudflare with DNS challange (it doesn't require any open ports, but require Cloudflare API token with `Zone.DNS` permission).
```yml
email: <SECRET>
domains:
  - <SECRET>
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-cloudflare
  cloudflare_api_token: <SECRET>
```
`Let's Encrypt` will create temporary DNS entry for challenging while recreating certificate in this configuration.

As a bonus point, you should create some scheduled job for recreating certificate. Renew process is run only on `Let's Encrypt` plugin start. [Example solution.](https://community.home-assistant.io/t/lets-encrypt-add-on-how-to-get-automatic-renewal/171236)

**Comment by rruizGit** [comment]
@sigo, thank you, sir!

**Comment by Veldkornet** [comment]
> I am also hitting this issue. What is happening is that the addon is requesting challenges for both the alias(es) as well as the domain(s). The second challenge is overwriting the first challenge before the challenge validation is taking place, thus resulting in a failed validation.
> 
> As a temporary workaround
> 
> 1. Remove all the aliases from the config and let it just validate the Duckdns domains. This completes successfully.
> 2. Put back the aliases into the config and let it re-validate all the domains. As the Duckdns domains are already validated, it will only deploy challenges to the Alias domains. This time it will complete successfully (if your alias domains point to different duckdns domains! In case more than one Alias point to the same Duckdns domain, you need to split step 2 into multiple steps and let it only validate one of the duplicate aliases at a time).
> 
> Edit:
> One downside of the current configuration design is that all the duckdns domains will be part of the certificate where previously I only had the domains in the Domain section as well as the Aliases on the certificate.

Thanks, this workaround works for me too. Hopefully I won’t need to do this every time though…

**Comment by marcomsousa** [comment]
**I found the root cause:**
The upgrade of de dependency dehydrated.

Since dehydrated 0.6.0, dehydrated change the domain validation strategy.
Until that version, they was validation in sequential, and change to validate in parallel.

Now they, deploy all TXT for all the domains, and validate all the domains. This fails because when dehydrated starts to validate   we only have the last TXT record in duckdns.org

dehydrated doesn't allow to change to the old strategy:
https://github.com/dehydrated-io/dehydrated/blob/master/docs/troubleshooting.md#dns-invalid-challenge-since-dehydrated-060--why-are-dns-challenges-deployed-first-and-verified-later

**Comment by rruizGit** [comment]
> Can we pin the requirement on `dehydrated` to [0.5.0](https://github.com/dehydrated-io/dehydrated/tree/v0.5.0)? I imagine there are security implications there.

Or, if DuckDNS will never work correctly in this situation, @sigo's solution might be the way to go?

**Comment by marcomsousa** [comment]
The problem is not DuckDNS, is that dehydrated now is incompatible with DuckDNS for multidomains alias.
The best aproche is to change from dehydrated to acme.sh (https://github.com/acmesh-official/acme.sh/wiki/DNS-alias-mode)

**Comment by sigo** [comment]
> > Can we pin the requirement on `dehydrated` to [0.5.0](https://github.com/dehydrated-io/dehydrated/tree/v0.5.0)? I imagine there are security implications there.
> 
> Or, if DuckDNS will never work correctly in this situation, @sigo's solution might be the way to go?

My semi-manual way works, but it is workaround. it may work better (via DuckDNS). I think @marcomsousa's research is very valuable here - and it doesn't seems like "I think i know the issue".

**Comment by marcomsousa** [comment]
![image](https://user-images.githubusercontent.com/1110889/130150275-8844a618-8773-4f08-9ace-59943cfb2e95.png)

I change the hook.sh to print some debug information, so I'm 100% sure.

1) Deploy TXT for ALIAS
2) Deploy TXT for DOMAIN
3) Checking TXT for ALIAS, failed becase is ALIAS TXT on that system.

It's easy to change to acme.sh or getssl (removing dehydrated) and complete fix this issue.

**Comment by Veldkornet** [comment]
Nope, still broken

**Comment by p0wertiger** [comment]
I hit this problem 3 days ago. Thanks for the manual workaround, did the job for now but the issue is still not fixed...

**Comment by sh3llc0de** [comment]
This add-on is majorly broken.   I did test and validate duckdns with the acme.sh script which works flawlessly with a wildcard certificate:

- export DuckDNS_Token="secret-token"
- acme.sh --issue -d '*.testdomain.us' -k ec-384 --dns dns_duckdns --challenge-alias testdomain.duckdns.org

You can install acme with:
- curl https://get.acme.sh | sh

And this just requires one CNAME:
_acme-challenge.testdomain.us -> testdomain.duckdns.org

You can then renew the certificate with:
- acme.sh --cron

If I knew how to modify the add-on to incorporate acme.sh I would supply a patch. 

What are the chances we can make this work?

**Comment by hjed** [comment]
I also have this issue

**Comment by pavelkryl** [comment]
The issue still persists with DuckDNS 1.14.0

**Comment by enjikaka** [comment]
Now my certs expired and DuckDNS 1.14.0 didn't renew because of the alias, once again... It has not worked a single time. Fix from @Veldkornet works still though.

Can't the plugin just automate the user pain here? Renew the certs in two steps where the first step ignores aliases?

**Comment by Veldkornet** [comment]
Nope Mr Bot, it's still buggered.

**Comment by fume** [comment]
i can confirm the issue is still present as of today: dns-01 challenge fails if you have aliases configured right from the first start of the extension.

**Comment by Veldkornet** [comment]
Still very much an issue

**Comment by matlimatli** [comment]
So based on the findings of @marcomsousa, it seems like we have three options for solving this:

1. Convince duckdns to add support for multiple TXT records per domain
2. Convince the dehydrated maintainer to add support for sequential updates (see for example the discussion in [https://github.com/dehydrated-io/dehydrated/pull/553](url))
3. Replace dehydrated with acme.sh

Any other ideas?

**Comment by GeekGoneOld** [comment]
> Any other ideas?

Is it not also possible to install DuckDNS with:

lets encrypt:
  accept_terms: false

AND install the let's encrypt add-in to handle the cert?  I read somewhere that let's encrypt add-in doesn't have this same problem.

**Comment by fume** [comment]
> So based on the findings of @marcomsousa, it seems like we have three options for solving this:
> 
> 1. Convince duckdns to add support for multiple TXT records per domain
> 2. Convince the dehydrated maintainer to add support for sequential updates (see for example the discussion in [https://github.com/dehydrated-io/dehydrated/pull/553](url))
> 3. Replace dehydrated with acme.sh
> 
> Any other ideas?

If i were the mantainer of this addon, I would go for n.3 so everything is under "my" control.

My2c

regards,
F.

**Comment by Veldkornet** [comment]
Still a problem

**Comment by kricha** [comment]
@ludeeus any news of this issue more than year old?

**Comment by ludeeus** [comment]
@kricha did you open a PR? I can't find that?

**Comment by kricha** [comment]
> @kricha did you open a PR? I can't find that?

??? i'm not developer of this small add-on. Maybe you could ping author of this?

**Comment by Veldkornet** [comment]
Still an issue in version 1.15.0

**Comment by Veldkornet** [comment]
Still an issue

**Comment by vanto** [comment]
Still an issue.

**Comment by hjed** [comment]
Still broken

**Comment by harryinet** [comment]
Yes, still broken and urgent.

**Comment by Oranginator** [comment]
> Yes, still broken and urgent.

Yeah I also have this issue.

**Comment by andreasc1** [comment]
I also have this issue.

**Comment by Oranginator** [comment]
> I also have this issue.


I had the following config when the problem occurred:

> domains:
>   - xxxxx.duckdns.org
> token: xxxx
> aliases: []
> lets_encrypt:
>   accept_terms: true
>   algo: secp384r1
>   certfile: fullchain.pem
>   keyfile: privkey.pem
> seconds: 300
> ipv4: https://api.ipify.org/
> ipv6: https://api6.ipify.org/

After removing those lines:

> ipv4: https://api.ipify.org/
> ipv6: https://api6.ipify.org/

It works now for me.

**Comment by himynameisjonas** [comment]
This is still broken 😢

**Comment by harryinet** [comment]
Mine works fine since weeks

**Comment by hjed** [comment]
Still broken for me

**Comment by p0wertiger** [comment]
Would be so nice to have this fixed finally

**Comment by Veldkornet** [comment]
Still broken

**Comment by p0wertiger** [comment]
Still not solved

**Comment by p0wertiger** [comment]
Not solved

**Comment by Veldkornet** [comment]
Still an issue, stupid bot 🤖

**Comment by Clusters** [comment]
Issue still persistent. Thanks for removing stale!

**Comment by Veldkornet** [comment]
Still an issue

**Comment by mansouryaacoubi** [comment]
I can confirm that my workaround from February 2021 is still working. But unfortunately there is no permanent fix. Maybe someone has to fork the DuckDNS repo and fix it for themselves if the current addon is unmaintained.

https://github.com/home-assistant/addons/issues/1331#issuecomment-780495261

**Comment by hjed** [comment]
still broken

**Comment by Xebozone** [comment]
@pvizeli
Related to https://github.com/home-assistant/addons/pull/1785 which you merged

**Comment by Xebozone** [comment]
bumping this. I don't think it's resolved

**Comment by dimmuboy** [comment]
> I can confirm that my workaround from February 2021 is still working. But unfortunately there is no permanent fix. Maybe someone has to fork the DuckDNS repo and fix it for themselves if the current addon is unmaintained.
> 
> [#1331 (comment)](https://github.com/home-assistant/addons/issues/1331#issuecomment-780495261)

it does not work anymore because addon is checking domains with regex `.+\.duckdns\.org`

**Comment by iranl** [comment]
#3152 added support for DuckDNS DNS challenges to the Let's Encrypt addon in version 5.0.1.

I would suggest using the DuckDNS only for keeping a dynamic IP up to date at DuckDNS and use the Let's Encrypt addon to get (wildcard) certs for custom domains using DuckDNS DNS challenges.

**Comment by hjed** [comment]
Hi @iranl I looked at that yesterday, but it looks like it would only support *.duckdns.org domains and not alias domains?

**Comment by iranl** [comment]
@hjed Also supports (wildcard) custom domains.

Example YAML config:

```
domains:
  - <(WILDCARD) CUSTOMDOMAIN>
email: <EMAIL>
keyfile: privkey.pem
certfile: fullchain.pem
challenge: dns
dns:
  propagation_seconds: 60
  provider: dns-duckdns
  duckdns_token: <DUCKDNSTOKEN>
```

CNAME for custom domain "test.example.com" with DuckDNS domain "abc.duckdns.org" should be set up as follows:

`_acme-challenge.test.example.com. 600 IN CNAME abc.duckdns.org.`

**Comment by marcomsousa** [comment]
Issue still active

**Comment by jum0n** [comment]
Issue was still a problem for me.  I removed all my alternates, restarted the add-on, and got a cert update. Added the alternates back and got another update successfully.  Restarted HA and cleared the cache on Firefox and it works now.

**Comment by hjed** [comment]
Still an issue. 

As a side note - I tried the workaround using the lets encrypt plugin suggested by @iranl but it appears that is also broken:

```
Encountered exception during recovery: certbot.errors.PluginError: The domain "_acme-challenge.<my domain>.duckdns.org" is not valid a duckdns subdomain.
The domain "_acme-challenge.<my domain>.duckdns.org" is not valid a duckdns subdomain.
```

So not a workaround that can be used. Still need to remove all the alternatives and re-add every time the certificate expires in the duckdns plugin.

**Comment by p0wertiger** [comment]
Still active

**Comment by p0wertiger** [comment]
Still not solved

**Comment by kricha** [comment]
3 years issue :D

**Comment by pyrech** [comment]
For the record, here is how I bypassed this issue  yesterday by using the Nginx Proxy Manager addon.

- first, I removed the alias domain on the duckdns configuration
- I defined a new CNAME entry on my DNS provider  (homeassistant.example.com. CNAME myproject.duckdns.org.)

From now on, duckdns' only mission is to handle the dynamic IP stuff and my custom domain will always points to my dynamic IP. Great, let's fixe the real SSL issue.

I gave a try to the "Let's Encrypt" addon to generate a certificate for my custom domain, then registering these certificates inside HA directly. It worked but it has 2 downsides for me:
- HA is now in https only. This means it can no longer be accessed in http, **even locally**
- certificates renewal is not available out of the box, you have to restart the addon periodically to trigger the renewal (it can be done with automation though)

For this reason, I used the "Nginx Proxy Manager" addon instead as it provides builtin support for let's encrypt certificates.
Once installed, I:
- changed the default admin credentials (check the addon readme)
- configured a new proxy with my custom domain homeassistant.example.com, to proxify traffic (and handle Let's Encrypt certificates) to the local IP and port (8123 by default) of home assistant, in http.
- in your router, open the port 443 (and 80 if you want http access from outside) and point it to the ip of HA but on the port 443 (and 80) : this the port(s) that the nginx addon listen on your device running HA.

You will need to configure HA to accept request from this proxy, so in the configuration.yaml, add the following:

```yaml
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 172.30.33.0/24
```

Everything should now works as expected when using HTTPS with your custom domain, if not, check the log from the addon and from HA.

On a side note, I also kept the port 8123 from my router open on the one from HA, as it this not my main house, I want a direct access to HA in case something gone wrong one day on nginx/ssl certificates/whatever 😝

**Comment by Veldkornet** [comment]
Still broken

**Comment by Veldkornet** [comment]
Still buggered

**Comment by Veldkornet** [comment]
Still broken...

**Comment by Veldkornet** [comment]
Still broken

**Comment by Xebozone** [comment]
> Still broken

Thanks for keeping the issue alive. I just wish they would fix it!

**Comment by Veldkornet** [comment]
Nope. It is, as always, still broken.

**Comment by Veldkornet** [comment]
It's busted as always Mr Bot

**Comment by Veldkornet** [comment]
> This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

No

**Comment by marcomsousa** [comment]
Still present

**Comment by p0wertiger** [comment]
Definitely not solved yet

**Comment by Xebozone** [comment]
NOT FIXED

**Comment by necrose99** [comment]
domains:
  - null
  - lo******.duckdns.org
  - ne****.duckdns.org
email: mike@michaellawrenceit.com
keyfile: privkey.pem
certfile: fullchain.pem
challenge: dns
dns: {dns-duckdns}
duckdns_email: mi*****
duckdns_api_key: ******9
key_type: rsa
elliptic_curve: secp384r1

duckdns is a legacy integration however via Lets encypt add on a plus.

**Comment by Xebozone** [comment]
Just pinging in advance... NOT SOLVED YET

**Comment by marcomsousa** [comment]
Finally I changed this addon with:
- a private domain in cloudflare,
- a cloudflare integration for ddns
- a let's encrypt addon with a automation trigger weekly.
- Also a optional nginx proxy addon for having HA without https in the internal Network.

**Comment by GoetzGoerisch** [comment]
Please verify with 5.3.3

**Comment by harryinet** [comment]
I use Dynu DNS because all these problems. This works well since months.

**Comment by Xebozone** [comment]
💩 not fixed

**Comment by Xebozone** [comment]
Still broken for me on my most recent renewal

**Comment by Veldkornet** [comment]
Still an issue

**Comment by Xebozone** [comment]
oh boo. Well. We kept it open since 2021...This issue had a good run.... still not fixed though.

**Comment by Veldkornet** [comment]
Indeed, we forgot to comment. 

I assume that the hope was that if the issue was ignored long enough, people (and thereby the issue) would just be forgotten.

**Comment by Xebozone** [comment]
Good news, maybe!
I got really tired of this bug, and I believe I've fixed it in https://github.com/home-assistant/addons/pull/4106
Waiting for my PR to hopefully get approved!

**Comment by Veldkornet** [comment]
I have 1.22 installed at the moment, but ever since 1.20 I've been getting the following error. Anyone else in this issue with the same?

```
# INFO: Using main config file /data/workdir/config
+ Account already registered!
[19:20:10] INFO: Starting DuckDNS...
[19:20:10] INFO: Renew certificate for domains: my-domain.duckdns.org and aliases: 
ha.my-domain.com
# INFO: Using main config file /data/workdir/config
Processing ha.my-domain.com
 + Checking domain name(s) of existing cert... changed!
 + Domain name(s) are not matching!
 + Names in old certificate: ha.my-domain.com my-domain.duckdns.org
 + Configured names: ha.my-domain.com
 + Forcing renew.
 + Checking expire date of existing cert...
 + Valid till Aug 28 21:00:39 2025 GMT (Less than 30 days). Renewing!
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 1 authorizations URLs from the CA
 + Handling authorization for ha.my-domain.com
 + 1 pending challenge(s)
 + Deploying challenge tokens...
Processing domain: ha.my-domain.com
OK + Responding to challenge for ha.my-domain.com authorization...
 + Challenge is valid!
 + Cleaning challenge tokens...
OK + Requesting certificate...
Warning: Will read cert request from stdin since no -in option is given
 + Checking certificate...
 + Done!
 + Creating fullchain.pem...
 + Done!
# INFO: Using main config file /data/workdir/config
Processing my-domain.duckdns.org
 + Checking domain name(s) of existing cert... unchanged.
 + Checking expire date of existing cert...
 + Valid till Aug 24 19:50:22 2025 GMT (Less than 30 days). Renewing!
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 1 authorizations URLs from the CA
 + Handling authorization for my-domain.duckdns.org
 + 1 pending challenge(s)
 + Deploying challenge tokens...
Processing domain: my-domain.duckdns.org
Terminated
ERROR: deploy_challenge hook returned with non-zero exit code
```

**Comment by Xebozone** [comment]
What's the TTL on your CNAMEs for your domain(s)? Try setting it to a low value, like '1'. Then wait an hour and try again.

**Comment by Veldkornet** [comment]
> What's the TTL on your CNAMEs for your domain(s)? Try setting it to a low value, like '1'. Then wait an hour and try again.

TTL was 60, but I changed it to 1 and indeed it seems to have renewed now. 
Except, when I look at the certificate after restarting nginx, the certificate seems to be only having the duckdns domain as an alternate name and is missing the alias domain. 

If I restart the duckdns add-on, it seems to think that all th certs are correct, so it doesn't do anything. 

So something seems to have gone wrong somewhere with mine, but overall it's looking better than what it was. 

I removed everything in the addon's /data/letsencrypt folder and restarted, and the same happened. Also if I check the letsencrypt directory, it does seems to have certain for both domains. Unfortunately I'm not able to check the files themselves at the moment, but maybe it's just computing the wrong certificate at the end and it's copying the one without the alias?

**Comment by Xebozone** [comment]
> > What's the TTL on your CNAMEs for your domain(s)? Try setting it to a low value, like '1'. Then wait an hour and try again.
> 
> TTL was 60, but I changed it to 1 and indeed it seems to have renewed now. Except, when I look at the certificate after restarting nginx, the certificate seems to be only having the duckdns domain as an alternate name and is missing the alias domain.
> 
> If I restart the duckdns add-on, it seems to think that all th certs are correct, so it doesn't do anything.
> 
> So something seems to have gone wrong somewhere with mine, but overall it's looking better than what it was.
> 
> I removed everything in the addon's /data/letsencrypt folder and restarted, and the same happened. Also if I check the letsencrypt directory, it does seems to have certain for both domains. Unfortunately I'm not able to check the files themselves at the moment, but maybe it's just computing the wrong certificate at the end and it's copying the one without the alias?

I had to restart my Home Assistant VM (or machine) to get my cert to apply. You could try that?

**Comment by Veldkornet** [comment]
> 
> I had to restart my Home Assistant VM (or machine) to get my cert to apply. You could try that?

Well, the DuckDNS is copying them to
```
http:
  ssl_certificate: /ssl/fullchain.pem
  ssl_key: /ssl/privkey.pem
```
Which is where the nginx is reading the files from, and if I restart the addon, I can see that it uses new certs. I can see on the certificate validity that the date changed. But still, no alias in the alternate names. 

I'm not using SSL on homeassistant itself, but rather in the nginx addon on port 443.

**Comment by Xebozone** [comment]
> > I had to restart my Home Assistant VM (or machine) to get my cert to apply. You could try that?
> 
> Well, the DuckDNS is copying them to
> 
> ```
> http:
>   ssl_certificate: /ssl/fullchain.pem
>   ssl_key: /ssl/privkey.pem
> ```
> 
> Which is where the nginx is reading the files from, and if I restart the addon, I can see that it uses new certs. I can see on the certificate validity that the date changed. But still, no alias in the alternate names.
> 
> I'm not using SSL on homeassistant itself, but rather in the nginx addon on port 443.

Thanks! That makes sense. I'll take a look at this later.
The changes I made tell the addon to process each domain or alias separately, but it sounds like it doesn't know to combine the certificates into a single certificate.

**Comment by Veldkornet** [comment]
> 
> Thanks! That makes sense. I'll take a look at this later.
> The changes I made tell the addon to process each domain or alias separately, but it sounds like it doesn't know to combine the certificates into a single certificate.

I don't think it should be combining certificates per se, it should just add the alias as a Subject Alternative Name (SAN) on the DuckDNS certificate. 

With dehydrated, I believe you just need to ensure that the alias name is added to the domains.txt file when the request is done. 

_domains.txt_
```
my-domain.duckdns.org ha.my-domain.com
```

You will get a certificate with:
	•	CN = my-domain.duckdns.org
	•	SANs = my-domain.duckdns.org, ha.my-domain.com

**Comment by Xebozone** [comment]
> > I had to restart my Home Assistant VM (or machine) to get my cert to apply. You could try that?
> 
> Well, the DuckDNS is copying them to
> 
> ```
> http:
>   ssl_certificate: /ssl/fullchain.pem
>   ssl_key: /ssl/privkey.pem
> ```
> 
> Which is where the nginx is reading the files from, and if I restart the addon, I can see that it uses new certs. I can see on the certificate validity that the date changed. But still, no alias in the alternate names.
> 
> I'm not using SSL on homeassistant itself, but rather in the nginx addon on port 443.

Hopefully finally fixed now in https://github.com/home-assistant/addons/pull/4112 (pending at this time)
Tested on my machine. I inspected the fullchain.pem and it contains the SANs

**Comment by Veldkornet** [comment]
> 
> Hopefully finally fixed now in https://github.com/home-assistant/addons/pull/4112 (pending at this time)
> Tested on my machine. I inspected the fullchain.pem and it contains the SANs

Okay, updated to 1.24. Looks promising, but seems to fail on the final step. I tried with just letting it restart, then afterwards by cleaning up the certs first. Both had the same result. 

Initial restart:
```

s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service duckdns: starting
s6-rc: info: service duckdns successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
# INFO: Using main config file /data/workdir/config
+ Account already registered!
[10:09:24] INFO: Starting DuckDNS...
[10:09:24] INFO: Renew certificate for domains: my-domain.duckdns.org and aliases: 
ha.my-domain.com
# INFO: Using main config file /data/workdir/config
Processing ha.my-domain.com
 + Checking domain name(s) of existing cert... unchanged.
 + Checking expire date of existing cert...
 + Valid till Nov  9 20:05:14 2025 GMT (Longer than 30 days). Skipping renew!
# INFO: Using main config file /data/workdir/config
Processing my-domain.duckdns.org
 + Checking domain name(s) of existing cert... unchanged.
 + Checking expire date of existing cert...
 + Valid till Nov  9 20:05:21 2025 GMT (Longer than 30 days). Skipping renew!
[10:09:27] INFO: Final pass to combine individual certificates
# INFO: Using main config file /data/workdir/config
Processing ha.my-domain.com with alternative names: my-domain.duckdns.org
 + Checking domain name(s) of existing cert... changed!
 + Domain name(s) are not matching!
 + Names in old certificate: ha.my-domain.com
 + Configured names: ha.my-domain.com my-domain.duckdns.org
 + Forcing renew.
 + Checking expire date of existing cert...
 + Valid till Nov  9 20:05:14 2025 GMT (Longer than 30 days). Ignoring because renew was forced!
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 2 authorizations URLs from the CA
 + Handling authorization for ha.my-domain.com
 + Handling authorization for my-domain.duckdns.org
 + 2 pending challenge(s)
 + Deploying challenge tokens...
Processing domain: ha.my-domain.com
OKProcessing domain: my-domain.duckdns.org
Terminated
ERROR: deploy_challenge hook returned with non-zero exit code
```

Cert cleanup first:
```

s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service duckdns: starting
s6-rc: info: service duckdns successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
# INFO: Using main config file /data/workdir/config
+ Account already registered!
[10:20:52] INFO: Starting DuckDNS...
[10:20:53] INFO: Renew certificate for domains: my-domain.duckdns.org and aliases: 
ha.my-domain.com
# INFO: Using main config file /data/workdir/config
Processing ha.my-domain.com
 + Creating new directory /data/letsencrypt/ha.my-domain.com ...
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 1 authorizations URLs from the CA
 + Handling authorization for ha.my-domain.com
 + Found valid authorization for ha.my-domain.com
 + 0 pending challenge(s)
 + Requesting certificate...
Warning: Will read cert request from stdin since no -in option is given
 + Checking certificate...
 + Done!
 + Creating fullchain.pem...
 + Done!
# INFO: Using main config file /data/workdir/config
Processing my-domain.duckdns.org
 + Creating new directory /data/letsencrypt/my-domain.duckdns.org ...
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 1 authorizations URLs from the CA
 + Handling authorization for my-domain.duckdns.org
 + Found valid authorization for my-domain.duckdns.org
 + 0 pending challenge(s)
 + Requesting certificate...
Warning: Will read cert request from stdin since no -in option is given
 + Checking certificate...
 + Done!
 + Creating fullchain.pem...
 + Done!
[10:21:14] INFO: Final pass to combine individual certificates
# INFO: Using main config file /data/workdir/config
Processing ha.my-domain.com with alternative names: my-domain.duckdns.org
 + Checking domain name(s) of existing cert... changed!
 + Domain name(s) are not matching!
 + Names in old certificate: ha.my-domain.com
 + Configured names: ha.my-domain.com my-domain.duckdns.org
 + Forcing renew.
 + Checking expire date of existing cert...
 + Valid till Nov 11 07:22:32 2025 GMT (Longer than 30 days). Ignoring because renew was forced!
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 2 authorizations URLs from the CA
 + Handling authorization for ha.my-domain.com
 + Handling authorization for my-domain.duckdns.org
 + 2 pending challenge(s)
 + Deploying challenge tokens...
Processing domain: ha.my-domain.com
OKProcessing domain: my-domain.duckdns.org
Terminated
ERROR: deploy_challenge hook returned with non-zero exit code
```

**Comment by Xebozone** [comment]
> > Hopefully finally fixed now in [#4112](https://github.com/home-assistant/addons/pull/4112) (pending at this time)
> > Tested on my machine. I inspected the fullchain.pem and it contains the SANs
> 
> Okay, updated to 1.24. Looks promising, but seems to fail on the final step. I tried with just letting it restart, then afterwards by cleaning up the certs first. Both had the same result.
> 
> Initial restart:
> 
> ```
> 
> s6-rc: info: service s6rc-oneshot-runner: starting
> s6-rc: info: service s6rc-oneshot-runner successfully started
> s6-rc: info: service fix-attrs: starting
> s6-rc: info: service fix-attrs successfully started
> s6-rc: info: service legacy-cont-init: starting
> s6-rc: info: service legacy-cont-init successfully started
> s6-rc: info: service duckdns: starting
> s6-rc: info: service duckdns successfully started
> s6-rc: info: service legacy-services: starting
> s6-rc: info: service legacy-services successfully started
> # INFO: Using main config file /data/workdir/config
> + Account already registered!
> [10:09:24] INFO: Starting DuckDNS...
> [10:09:24] INFO: Renew certificate for domains: my-domain.duckdns.org and aliases: 
> ha.my-domain.com
> # INFO: Using main config file /data/workdir/config
> Processing ha.my-domain.com
>  + Checking domain name(s) of existing cert... unchanged.
>  + Checking expire date of existing cert...
>  + Valid till Nov  9 20:05:14 2025 GMT (Longer than 30 days). Skipping renew!
> # INFO: Using main config file /data/workdir/config
> Processing my-domain.duckdns.org
>  + Checking domain name(s) of existing cert... unchanged.
>  + Checking expire date of existing cert...
>  + Valid till Nov  9 20:05:21 2025 GMT (Longer than 30 days). Skipping renew!
> [10:09:27] INFO: Final pass to combine individual certificates
> # INFO: Using main config file /data/workdir/config
> Processing ha.my-domain.com with alternative names: my-domain.duckdns.org
>  + Checking domain name(s) of existing cert... changed!
>  + Domain name(s) are not matching!
>  + Names in old certificate: ha.my-domain.com
>  + Configured names: ha.my-domain.com my-domain.duckdns.org
>  + Forcing renew.
>  + Checking expire date of existing cert...
>  + Valid till Nov  9 20:05:14 2025 GMT (Longer than 30 days). Ignoring because renew was forced!
>  + Signing domains...
>  + Generating private key...
>  + Generating signing request...
>  + Requesting new certificate order from CA...
>  + Received 2 authorizations URLs from the CA
>  + Handling authorization for ha.my-domain.com
>  + Handling authorization for my-domain.duckdns.org
>  + 2 pending challenge(s)
>  + Deploying challenge tokens...
> Processing domain: ha.my-domain.com
> OKProcessing domain: my-domain.duckdns.org
> Terminated
> ERROR: deploy_challenge hook returned with non-zero exit code
> ```
> 
> Cert cleanup first:
> 
> ```
> 
> s6-rc: info: service s6rc-oneshot-runner: starting
> s6-rc: info: service s6rc-oneshot-runner successfully started
> s6-rc: info: service fix-attrs: starting
> s6-rc: info: service fix-attrs successfully started
> s6-rc: info: service legacy-cont-init: starting
> s6-rc: info: service legacy-cont-init successfully started
> s6-rc: info: service duckdns: starting
> s6-rc: info: service duckdns successfully started
> s6-rc: info: service legacy-services: starting
> s6-rc: info: service legacy-services successfully started
> # INFO: Using main config file /data/workdir/config
> + Account already registered!
> [10:20:52] INFO: Starting DuckDNS...
> [10:20:53] INFO: Renew certificate for domains: my-domain.duckdns.org and aliases: 
> ha.my-domain.com
> # INFO: Using main config file /data/workdir/config
> Processing ha.my-domain.com
>  + Creating new directory /data/letsencrypt/ha.my-domain.com ...
>  + Signing domains...
>  + Generating private key...
>  + Generating signing request...
>  + Requesting new certificate order from CA...
>  + Received 1 authorizations URLs from the CA
>  + Handling authorization for ha.my-domain.com
>  + Found valid authorization for ha.my-domain.com
>  + 0 pending challenge(s)
>  + Requesting certificate...
> Warning: Will read cert request from stdin since no -in option is given
>  + Checking certificate...
>  + Done!
>  + Creating fullchain.pem...
>  + Done!
> # INFO: Using main config file /data/workdir/config
> Processing my-domain.duckdns.org
>  + Creating new directory /data/letsencrypt/my-domain.duckdns.org ...
>  + Signing domains...
>  + Generating private key...
>  + Generating signing request...
>  + Requesting new certificate order from CA...
>  + Received 1 authorizations URLs from the CA
>  + Handling authorization for my-domain.duckdns.org
>  + Found valid authorization for my-domain.duckdns.org
>  + 0 pending challenge(s)
>  + Requesting certificate...
> Warning: Will read cert request from stdin since no -in option is given
>  + Checking certificate...
>  + Done!
>  + Creating fullchain.pem...
>  + Done!
> [10:21:14] INFO: Final pass to combine individual certificates
> # INFO: Using main config file /data/workdir/config
> Processing ha.my-domain.com with alternative names: my-domain.duckdns.org
>  + Checking domain name(s) of existing cert... changed!
>  + Domain name(s) are not matching!
>  + Names in old certificate: ha.my-domain.com
>  + Configured names: ha.my-domain.com my-domain.duckdns.org
>  + Forcing renew.
>  + Checking expire date of existing cert...
>  + Valid till Nov 11 07:22:32 2025 GMT (Longer than 30 days). Ignoring because renew was forced!
>  + Signing domains...
>  + Generating private key...
>  + Generating signing request...
>  + Requesting new certificate order from CA...
>  + Received 2 authorizations URLs from the CA
>  + Handling authorization for ha.my-domain.com
>  + Handling authorization for my-domain.duckdns.org
>  + 2 pending challenge(s)
>  + Deploying challenge tokens...
> Processing domain: ha.my-domain.com
> OKProcessing domain: my-domain.duckdns.org
> Terminated
> ERROR: deploy_challenge hook returned with non-zero exit code
> ```

Useful. Thanks!
Would you mind sharing your full DuckDNS config?
Be sure to remove the token and to obfuscate your private domains.

**Comment by Veldkornet** [comment]
> 
> Useful. Thanks!
> Would you mind sharing your full DuckDNS config?
> Be sure to remove the token and to obfuscate your private domains.
```
domains:
  - my-domain.duckdns.org
token: xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx
aliases:
  - domain: ha.my-domain.com
    alias: my-domain.duckdns.org
lets_encrypt:
  accept_terms: true
  algo: secp384r1
  certfile: fullchain.pem
  keyfile: privkey.pem
seconds: 300
```

**Comment by Veldkornet** [comment]
Seems to be working as expected in 1.25.0. 

Certificate renewed and has the alias as an alternative name 🥳🥳🥳

**Comment by Xebozone** [comment]
> Seems to be working as expected in 1.25.0.
> 
> Certificate renewed and has the alias as an alternative name 🥳🥳🥳

Awesome! Thanks for your patience while I fixed things.
There's still one more teething issue I need to fix but it should be reliable now for most people

**Comment by Veldkornet** [comment]
Still working on 1.26.0 💪

**Comment by Xebozone** [comment]
Thanks for the feedback! Glad it's working. 💪
